src/entity/Living.php

```diff
@@ -920,4 +924,70 @@ protected function destroyCycles() : void{
 		);
 		parent::destroyCycles();
 	}
+
+	protected function blockedByShield(EntityDamageEvent $source) : bool{
+		$damager = null;
+		if($source instanceof EntityDamageByChildEntityEvent){
+			$damager = $source->getChild();
+		}elseif($source instanceof EntityDamageByEntityEvent){
+			$damager = $source->getDamager();
+		}
+
+		if($damager === null || !$this->isBlocking()){
+			return false;
+		}
+
+		$entityPos = $damager->getPosition();
+		$direction = $this->getDirectionVector();
+		$normalizedVector = $this->getPosition()->subtractVector($entityPos)->normalize();
+
+		$blocked = ($normalizedVector->x * $direction->x) + ($normalizedVector->z * $direction->z) < 0.0;
+
+		$event = new EntityDamageBlockedEvent($damager, $this, $source->getFinalDamage());
+
+		if(!$blocked || !$source->canBeReducedByArmor()){
+			$event->cancel();
+		}
+
+		if($event->isCancelled()){
+			return false;
+		}
+
+		if($damager instanceof Living){
+			$deltaX = $damager->getPosition()->x - $this->getPosition()->x;
+			$deltaZ = $damager->getPosition()->z - $this->getPosition()->z;
+			$damager->knockBack($deltaX, $deltaZ);
+			$damager->attackTime = 10;
+			$damager->isAttackTimeByShieldKb = true;
```

Comment:
Why do we need to reset the damager's invincibility ticks and flag if it was by shield? This doesn't seem necessary